torchao/quantization/config_parser.py

```diff
@@ -0,0 +1,226 @@
+ConfigType = Type[AOBaseConfig]
+
+
+# Define a Protocol for parameter processors
+class ParameterProcessor(Protocol):
+    """Protocol defining the interface for parameter processors"""
+
+    def __call__(self, match: re.Match, quant_config: ConfigType) -> Tuple[str, Any]:
+        """
+        Process a regex match into a parameter name and value
+
+        Args:
+            match: The regex match object containing captured groups
+            quant_config: The quantization config class being instantiated
+
+        Returns:
+            Tuple of (parameter_name, parameter_value)
+
+        Note:
+            If you need special handling based on the quant_config type,
+            be sure to use issubclass instead of isinstance.
+        """
+        ...
+
+
+def process_bits(match: re.Match, quant_config: AOBaseConfig) -> Tuple[str, Any]:
+    return "bits", int(match.group(1))
+
+
+def process_group_size(match: re.Match, quant_config: AOBaseConfig) -> Tuple[str, Any]:
+    return "group_size", int(match.group(1))
+
+
+def process_activation_bits(
+    match: re.Match, quant_config: AOBaseConfig
+) -> Tuple[str, Any]:
+    return "activation_bits", int(match.group(1))
+
+
+def process_weight_bits(match: re.Match, quant_config: AOBaseConfig) -> Tuple[str, Any]:
+    return "weight_bits", int(match.group(1))
+
+
+def process_symmetry(match: re.Match, quant_config: AOBaseConfig) -> Tuple[str, Any]:
+    mapping_type = (
+        MappingType.SYMMETRIC if match.group(1) == "sym" else MappingType.ASYMMETRIC
+    )
+    return "mapping_type", mapping_type
+
+
+def process_dtype(match: re.Match, quant_config: AOBaseConfig) -> Tuple[str, Any]:
+    dtype_map = {
+        "int4": torch.int4,
+        "int8": torch.int8,
+        "uint4": torch.uint4,
+        "uint8": torch.uint8,
+        "e4m3": torch.float8_e4m3fn,
+        "e5m2": torch.float8_e5m2,
+    }
+    # The float8's have different key names :(
+    key = (
+        "weight_dtype"
+        if issubclass(
+            quant_config,
+            (
+                Float8WeightOnlyConfig,
+                Float8DynamicActivationFloat8WeightConfig,
+            ),
+        )
+        else "dtype"
+    )
+    return key, dtype_map[match.group(1)]
+
+
+def process_per_row(match: re.Match, quant_config: AOBaseConfig) -> Tuple[str, Any]:
+    return "per_row", True
+
+
+class ConfigParser:
+    """Parser for string-based configuration patterns"""
+
+    # Parameter patterns with their processing functions
+    param_patterns: Dict[re.Pattern, ParameterProcessor] = {
+        re.compile(r"(\d+)bit"): process_bits,
+        re.compile(r"g(\d+)"): process_group_size,
+        # re.compile(r"act(\d+)"): process_activation_bits,
+        # re.compile(r"w(\d+)"): process_weight_bits,
+        re.compile(r"(sym|asym)"): process_symmetry,
+        re.compile(r"(int4|int8|uint4|uint8|e4m3|e5m2)"): process_dtype,
+        re.compile(r"(per_row)"): process_per_row,
```

Comment:
thanks, one more thing is can you add layout here as well: https://github.com/pytorch/ao/blob/f4786920746713be10427840b581abfed26aabf3/torchao/quantization/quant_api.py#L814 this will help with https://github.com/huggingface/transformers/pull/36146